Flaky-test: PulsarFunctionsJavaThreadTest.testTumblingCountWindowTest

### Search before reporting

- [x] I searched in the [issues](https://github.com/apache/pulsar/issues) and found nothing similar.


### Example failure

https://github.com/apache/pulsar/actions/runs/23244679795/job/67572420863#step:12:5542

### Exception stacktrace

```text
[ERROR] org.apache.pulsar.tests.integration.functions.java.PulsarFunctionsJavaThreadTest.testTumblingCountWindowTest -- Time elapsed: 15.78 s <<< FAILURE!
java.lang.AssertionError: expected [1] but found [3]
	at org.testng.Assert.fail(Assert.java:110)
	at org.testng.Assert.failNotEquals(Assert.java:1577)
	at org.testng.Assert.assertEqualsImpl(Assert.java:149)
	at org.testng.Assert.assertEquals(Assert.java:131)
	at org.testng.Assert.assertEquals(Assert.java:979)
	at org.testng.Assert.assertEquals(Assert.java:955)
	at org.testng.Assert.assertEquals(Assert.java:989)
	at org.apache.pulsar.tests.integration.functions.PulsarFunctionsTest.testWindowFunction(PulsarFunctionsTest.java:338)
	at org.apache.pulsar.tests.integration.functions.java.PulsarFunctionsJavaTest.testTumblingCountWindowTest(PulsarFunctionsJavaTest.java:204)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.testng.internal.invokers.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:139)
	at org.testng.internal.invokers.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:47)
	at org.testng.internal.invokers.InvokeMethodRunnable.call(InvokeMethodRunnable.java:76)
	at org.testng.internal.invokers.InvokeMethodRunnable.call(InvokeMethodRunnable.java:11)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
```

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR!

--- a/tests/integration/src/test/java/org/apache/pulsar/tests/integration/functions/PulsarFunctionsTest.java
+++ b/tests/integration/src/test/java/org/apache/pulsar/tests/integration/functions/PulsarFunctionsTest.java
@@ -330,7 +330,7 @@ public abstract class PulsarFunctionsTest extends PulsarFunctionsTestBase {
         for (int i = 0; i < 3; i++) {
             producer.send(String.format("%d", i).getBytes());
         }
-        awaitAndVerifySubscriptionStats(inputTopicName, "public/default/" + functionName, 3, 3);
+        awaitAndVerifySubscriptionStats(inputTopicName, functionName, 3, 3);
 
         for (int i = 3; i < numOfMessages; i++) {
             producer.send(String.format("%d", i).getBytes());
@@ -359,23 +359,30 @@ public abstract class PulsarFunctionsTest extends PulsarFunctionsTestBase {
         assertThat(i).isGreaterThanOrEqualTo(expectedResults.length - 1);
 
         // test that all messages are acked
-        awaitAndVerifySubscriptionStats(inputTopicName, "public/default/" + functionName, 0, 0);
+        awaitAndVerifySubscriptionStats(inputTopicName, functionName, 0, 0);
 
         deleteFunction(functionName);
 
         getFunctionInfoNotFound(functionName);
     }
 
-    private void awaitAndVerifySubscriptionStats(String topicName, String subscriptionName, int expectedBacklog,
+    private void awaitAndVerifySubscriptionStats(String inputTopicName, String functionName, int expectedBacklog,
                                                  int expectedUnacked) {
-        Awaitility.await().atMost(10, TimeUnit.SECONDS).untilAsserted(() -> {
-            TopicStats stats = pulsarAdmin.topics().getStats(topicName, true);
-            SubscriptionStats subStats = stats.getSubscriptions().get(subscriptionName);
-
-            assertNotNull(subStats, "Subscription stats should not be null for: " + subscriptionName);
-            assertEquals(subStats.getMsgBacklog(), expectedBacklog, "Message backlog mismatch");
-            assertEquals(subStats.getUnackedMessages(), expectedUnacked, "Unacked messages mismatch");
-        });
+        Awaitility.await()
+                .pollDelay(Duration.ZERO)
+                // Check every 50ms
+                .pollInterval(Duration.ofMillis(50))
+                .atMost(Duration.ofSeconds(10))
+                .ignoreExceptions()
+                .untilAsserted(() -> {
+                    TopicStats currentStats = pulsarAdmin.topics().getStats(inputTopicName, true);
+                    SubscriptionStats currentSubStats =
+                            currentStats.getSubscriptions().get("public/default/" + functionName);
+                    assertNotNull(currentSubStats);
+                    // Compare actual to expected
+                    assertEquals(currentSubStats.getMsgBacklog(), expectedBacklog);
+                    assertEquals(currentSubStats.getUnackedMessages(), expectedUnacked);
+                });
     }
 
     protected void testFunctionNegAck(Runtime runtime) throws Exception {

--- a/tests/integration/src/test/java/org/apache/pulsar/tests/integration/functions/PulsarFunctionsTest.java
+++ b/tests/integration/src/test/java/org/apache/pulsar/tests/integration/functions/PulsarFunctionsTest.java
@@ -330,13 +330,7 @@ public abstract class PulsarFunctionsTest extends PulsarFunctionsTestBase {
         for (int i = 0; i < 3; i++) {
             producer.send(String.format("%d", i).getBytes());
         }
-        Awaitility.await().atMost(10, TimeUnit.SECONDS).untilAsserted(() -> {
-            TopicStats currentStats = pulsarAdmin.topics().getStats(inputTopicName, true);
-            SubscriptionStats currentSubStats = currentStats.getSubscriptions().get("public/default/" + functionName);
-            assertNotNull(currentSubStats);
-            assertEquals(currentSubStats.getMsgBacklog(), 3);
-            assertEquals(currentSubStats.getUnackedMessages(), 3);
-        });
+        awaitAndVerifySubscriptionStats(inputTopicName, "public/default/" + functionName, 3, 3);
 
         for (int i = 3; i < numOfMessages; i++) {
             producer.send(String.format("%d", i).getBytes());
@@ -365,17 +359,23 @@ public abstract class PulsarFunctionsTest extends PulsarFunctionsTestBase {
         assertThat(i).isGreaterThanOrEqualTo(expectedResults.length - 1);
 
         // test that all messages are acked
-        Awaitility.await().atMost(10, TimeUnit.SECONDS).untilAsserted(() -> {
-            TopicStats finalStats = pulsarAdmin.topics().getStats(inputTopicName, true);
-            SubscriptionStats finalSubStats = finalStats.getSubscriptions().get("public/default/" + functionName);
-            assertNotNull(finalSubStats);
-            assertEquals(finalSubStats.getMsgBacklog(), 0);
-            assertEquals(finalSubStats.getUnackedMessages(), 0);
-        });
+        awaitAndVerifySubscriptionStats(inputTopicName, "public/default/" + functionName, 0, 0);
 
         deleteFunction(functionName);
 
         getFunctionInfoNotFound(functionName);
+    }
+
+    private void awaitAndVerifySubscriptionStats(String topicName, String subscriptionName, int expectedBacklog,
+                                                 int expectedUnacked) {
+        Awaitility.await().atMost(10, TimeUnit.SECONDS).untilAsserted(() -> {
+            TopicStats stats = pulsarAdmin.topics().getStats(topicName, true);
+            SubscriptionStats subStats = stats.getSubscriptions().get(subscriptionName);
+
+            assertNotNull(subStats, "Subscription stats should not be null for: " + subscriptionName);
+            assertEquals(subStats.getMsgBacklog(), expectedBacklog, "Message backlog mismatch");
+            assertEquals(subStats.getUnackedMessages(), expectedUnacked, "Unacked messages mismatch");
+        });
     }
 
     protected void testFunctionNegAck(Runtime runtime) throws Exception {

--- a/tests/integration/src/test/java/org/apache/pulsar/tests/integration/functions/PulsarFunctionsTest.java
+++ b/tests/integration/src/test/java/org/apache/pulsar/tests/integration/functions/PulsarFunctionsTest.java
@@ -330,11 +330,13 @@ public abstract class PulsarFunctionsTest extends PulsarFunctionsTestBase {
         for (int i = 0; i < 3; i++) {
             producer.send(String.format("%d", i).getBytes());
         }
-        TopicStats stats = pulsarAdmin.topics().getStats(inputTopicName, true);
-        SubscriptionStats subStats = stats.getSubscriptions().get("public/default/" + functionName);
-        assertNotNull(subStats);
-        assertEquals(3, subStats.getMsgBacklog());
-        assertEquals(3, subStats.getUnackedMessages());
+        Awaitility.await().atMost(10, TimeUnit.SECONDS).untilAsserted(() -> {
+            TopicStats currentStats = pulsarAdmin.topics().getStats(inputTopicName, true);
+            SubscriptionStats currentSubStats = currentStats.getSubscriptions().get("public/default/" + functionName);
+            assertNotNull(currentSubStats);
+            assertEquals(currentSubStats.getMsgBacklog(), 3);
+            assertEquals(currentSubStats.getUnackedMessages(), 3);
+        });
 
         for (int i = 3; i < numOfMessages; i++) {
             producer.send(String.format("%d", i).getBytes());
@@ -363,11 +365,13 @@ public abstract class PulsarFunctionsTest extends PulsarFunctionsTestBase {
         assertThat(i).isGreaterThanOrEqualTo(expectedResults.length - 1);
 
         // test that all messages are acked
-        stats = pulsarAdmin.topics().getStats(inputTopicName, true);
-        subStats = stats.getSubscriptions().get("public/default/" + functionName);
-        assertNotNull(subStats);
-        assertEquals(0, subStats.getMsgBacklog());
-        assertEquals(0, subStats.getUnackedMessages());
+        Awaitility.await().atMost(10, TimeUnit.SECONDS).untilAsserted(() -> {
+            TopicStats finalStats = pulsarAdmin.topics().getStats(inputTopicName, true);
+            SubscriptionStats finalSubStats = finalStats.getSubscriptions().get("public/default/" + functionName);
+            assertNotNull(finalSubStats);
+            assertEquals(finalSubStats.getMsgBacklog(), 0);
+            assertEquals(finalSubStats.getUnackedMessages(), 0);
+        });
 
         deleteFunction(functionName);
 

--- a/tests/integration/src/test/java/org/apache/pulsar/tests/integration/functions/PulsarFunctionsTest.java
+++ b/tests/integration/src/test/java/org/apache/pulsar/tests/integration/functions/PulsarFunctionsTest.java
@@ -369,7 +369,6 @@ public abstract class PulsarFunctionsTest extends PulsarFunctionsTestBase {
     private void awaitAndVerifySubscriptionStats(String inputTopicName, String functionName, int expectedBacklog,
                                                  int expectedUnacked) {
         Awaitility.await()
-                .pollDelay(Duration.ZERO)
                 // Check every 50ms
                 .pollInterval(Duration.ofMillis(50))
                 .atMost(Duration.ofSeconds(10))

--- a/tests/integration/src/test/java/org/apache/pulsar/tests/integration/functions/PulsarFunctionsTest.java
+++ b/tests/integration/src/test/java/org/apache/pulsar/tests/integration/functions/PulsarFunctionsTest.java
@@ -369,9 +369,6 @@ public abstract class PulsarFunctionsTest extends PulsarFunctionsTestBase {
     private void awaitAndVerifySubscriptionStats(String inputTopicName, String functionName, int expectedBacklog,
                                                  int expectedUnacked) {
         Awaitility.await()
-                // Check every 50ms
-                .pollInterval(Duration.ofMillis(50))
-                .atMost(Duration.ofSeconds(10))
                 .ignoreExceptions()
                 .untilAsserted(() -> {
                     TopicStats currentStats = pulsarAdmin.topics().getStats(inputTopicName, true);

